### PR TITLE
feature/SOF 7394

### PR DESCRIPTION
**Pull request by VsevolodX**
- feat: add island builder
- update: add island config
- update: add enum
- update: fix a mutable paramtere
- update: to cart and cryst
- update: cleanup
- update: remove
- update: gitignore
- chore: docstring
- update: pass any condition
- chore: fix enum
- update: add condition
- chore: lint fix
- update: make work
- chore: lint fix
- update: add coordinate condition class
- update: use coordcond class:
- update: adjsut other funcs
- chore: lint fix
- update: add create defect func
- update: add test
- update: add sphere
- try: adjust tests
- try: adjust tests
- try: add log
- try: add log
- try: add log 2
- try: add log 3
- try: add log 4
- try: add log 4
- try: remove supercell
- try: remove supercell 2
- try: C
- try: Ni
- try: Si json
- try: Si json
- update: optimize methods
- revert
- update: test with condiytion builder
- update: tests use 1x1 supercell
- try: move center
- update: use wheel server from utils
- try: ssh -> https
- feat: add terrace config
- update: adjust parameters
- update: add terrace builder
- feat: add filter by plane
- update: adjust to make work
- update: cleanup parameters
- chore: lint fix
- update: small adjustments
- update: rotate the atoms to meet pbc
- update: add roate function
- chore: fix error
- update: math

### CODE REVIEW

File: src/py/mat3ra/made/tools/build/defect/builders.py [hunk 1]
```diff
@@ -371,3 +378,143 @@ def _generate(self, configuration: _ConfigurationType) -> List[_GeneratedItemTyp
             thickness=configuration.thickness,
             use_cartesian_coordinates=configuration.use_cartesian_coordinates,
         )
+
+
+class TerraceSlabDefectBuilder(SlabDefectBuilder):
+    _ConfigurationType: type(TerraceSlabDefectConfiguration) = TerraceSlabDefectConfiguration  # type: ignore
+    _GeneratedItemType: Material = Material
+
+    def _calculate_cut_direction_vector(self, material, cut_direction):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/builders.py:387]: types, docstring for this and other functions

File: src/py/mat3ra/made/tools/build/defect/builders.py [hunk 2]
```diff
@@ -371,3 +378,143 @@ def _generate(self, configuration: _ConfigurationType) -> List[_GeneratedItemTyp
...
+        np_cut_direction = np.array(cut_direction)
+        direction_vector = np.dot(np.array(material.basis.cell.vectors_as_nested_array), np_cut_direction)
+        normalized_direction_vector = direction_vector / np.linalg.norm(direction_vector)
+        return normalized_direction_vector
+
+    def _calcualte_height(self, material, new_material):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/builders.py:393]: typo

File: src/py/mat3ra/made/tools/build/defect/builders.py [hunk 3]
```diff
@@ -371,3 +378,143 @@ def _generate(self, configuration: _ConfigurationType) -> List[_GeneratedItemTyp
...
+        hypotenuse = np.linalg.norm([height_cart, cut_direction_xy_proj_cart])
+        angle = np.arctan(height_cart / cut_direction_xy_proj_cart) * 180 / np.pi
+        normalized_rotation_axis = np.cross(normalized_direction_vector, [0, 0, 1]).tolist()
+        delta_length = hypotenuse - cut_direction_xy_proj_cart
+        return angle, normalized_rotation_axis, delta_length
+
+    def _increase_lattice_size(self, material, length_increase, direction_of_increase):
+        """Adjust the lattice vectors to accommodate the new structure dimensions."""
+        vector_a, vector_b = np.array(material.basis.cell.vector1), np.array(material.basis.cell.vector2)
+        delta_a_cart = np.dot(vector_a, direction_of_increase) * length_increase / np.linalg.norm(vector_a)
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/builders.py:415]: cart -> cartesian?

File: src/py/mat3ra/made/tools/build/defect/builders.py [hunk 4]
```diff
@@ -371,3 +378,143 @@ def _generate(self, configuration: _ConfigurationType) -> List[_GeneratedItemTyp
...
+        )
+        # Slope of the terrace along the cut direction
+        hypotenuse = np.linalg.norm([height_cart, cut_direction_xy_proj_cart])
+        angle = np.arctan(height_cart / cut_direction_xy_proj_cart) * 180 / np.pi
+        normalized_rotation_axis = np.cross(normalized_direction_vector, [0, 0, 1]).tolist()
+        delta_length = hypotenuse - cut_direction_xy_proj_cart
+        return angle, normalized_rotation_axis, delta_length
+
+    def _increase_lattice_size(self, material, length_increase, direction_of_increase):
+        """Adjust the lattice vectors to accommodate the new structure dimensions."""
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/builders.py:413]: Not sure what this does - needs a better explanation and an example in docstring maybe

File: src/py/mat3ra/made/tools/build/defect/builders.py [hunk 5]
```diff
@@ -371,3 +378,143 @@ def _generate(self, configuration: _ConfigurationType) -> List[_GeneratedItemTyp
...
+        material.lattice = new_lattice
+        return material
+
+    def _update_material_name(self, material, configuration):
+        updated_material = super()._update_material_name(material, configuration)
+        new_name = f"{updated_material.name}, {configuration.steps_number}-step Terrace {configuration.cut_direction}"
+        updated_material.name = new_name
+        return updated_material
+
+    def _rotate_material(self, material: Material, axis: List[int], angle: float) -> Material:
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/builders.py:438]: This is a generic function and can go to `modify` for material tools

File: src/py/mat3ra/made/tools/build/defect/builders.py [hunk 6]
```diff
@@ -371,3 +378,143 @@ def _generate(self, configuration: _ConfigurationType) -> List[_GeneratedItemTyp
...
+        if cut_direction is None:
+            cut_direction = [0, 0, 1]
+        if pivot_coordinate is None:
+            pivot_coordinate = [0.5, 0.5, 0.5]
+
+        new_material = material.clone()
+        material_with_additional_layers = self.create_material_with_additional_layers(new_material, steps_number)
+
+        normalized_direction_vector = self._calculate_cut_direction_vector(material, cut_direction)
+        condition = CoordinateConditionBuilder().plane(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/builders.py:490]: Why do we need to initiate this Builder? Should be `CoordinateConditionHolder` and `CoordinateConditionHolder.plane(...)`

File: src/py/mat3ra/made/tools/build/defect/builders.py [hunk 7]
```diff
@@ -371,3 +378,143 @@ def _generate(self, configuration: _ConfigurationType) -> List[_GeneratedItemTyp
...
+            plane_normal=normalized_direction_vector,
+            plane_point_coordinate=pivot_coordinate,
+        )[0]
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/builders.py:493]: What is the `[0]` for? It's ambiguous

File: src/py/mat3ra/made/tools/build/defect/builders.py [hunk 8]
```diff
@@ -371,3 +378,143 @@ def _generate(self, configuration: _ConfigurationType) -> List[_GeneratedItemTyp
...
+
+        if rotate_to_match_pbc:
+            adjusted_material = self._increase_lattice_size(result_material, delta_length, normalized_direction_vector)
+            result_material = self._rotate_material(
+                material=adjusted_material, axis=normalized_rotation_axis, angle=angle
+            )
+        return [result_material]
+
+    def _generate(self, configuration: _ConfigurationType) -> List[_GeneratedItemType]:
+        return self.create_terrace(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/builders.py:514]: self.create_terrace should return a single material and here you can create a 1-element array instead. Same for `create_island`

File: src/py/mat3ra/made/tools/build/defect/configuration.py [hunk 9]
```diff
@@ -163,3 +163,40 @@ def _json(self):
...
+        where the normal of the cut plane passes through.
+        steps_number (int): The number of steps to bunch (number of added layers).
+        use_cartesian_coordinates (bool): The flag to use cartesian coordinates for coordinates and vectors.
+        rotate_to_match_pbc (bool): The flag to rotate the slab with a terrace to match periodic boundary conditions.
+    """
+
+    defect_type: SlabDefectTypeEnum = SlabDefectTypeEnum.TERRACE
+    cut_direction: List[int] = [1, 0, 0]
+    pivot_coordinate: List[float] = [0.5, 0.5, 0.5]
+    steps_number: int = 1  # number of steps to bunch
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/configuration.py:187]: Let's just use `number of added layers`

File: src/py/mat3ra/made/tools/build/defect/configuration.py [hunk 10]
```diff
@@ -163,3 +163,40 @@ def _json(self):
...
+
+    @property
+    def _json(self):
+        return {
+            **super()._json,
+            "type": "TerraceSlabDefectConfiguration",
+            "defect_type": self.defect_type.name,
+            "cut_direction": self.cut_direction,
+            "pivot_coordinate": self.pivot_coordinate,
+            "steps_number": self.steps_number,
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/configuration.py:199]: Should be reused from SlabDefect

File: src/py/mat3ra/made/tools/build/defect/builders.py (hunk 5 above)
Comment (VsevolodX) [reply src/py/mat3ra/made/tools/build/defect/builders.py:438]: It means I need to add test for this too

File: src/py/mat3ra/made/tools/build/defect/configuration.py (hunk 10 above)
Comment (VsevolodX) [reply src/py/mat3ra/made/tools/build/defect/configuration.py]: I don't see similarities to inherit...

File: tests/py/unit/test_tools_build_defect.py [hunk 11]
```diff
@@ -128,3 +128,25 @@ def test_create_island():
...
+        use_orthogonal_z=True,
+    )
+    t = get_terminations(slab_config)[0]
+    slab = create_slab(slab_config, t)
+
+    config = TerraceSlabDefectConfiguration(
+        crystal=slab,
+        cut_direction=[1, 0, 0],
+        pivot_coordinate=[0.5, 0.5, 0.5],
+        steps_number=1,
```
Comment (timurbazhirov) [inline tests/py/unit/test_tools_build_defect.py:149]: This kwarg should not be there

File: src/py/mat3ra/made/tools/build/defect/configuration.py [hunk 12]
```diff
@@ -146,20 +162,53 @@ class IslandSlabDefectConfiguration(SlabDefectConfiguration):
...
+
+class TerraceSlabDefectConfiguration(SlabDefectConfiguration):
+    """
+    Configuration for a terrace slab defect.
+
+    Args:
+        crystal (Material): The Material object (must be a created slab).
+        defect_type (SlabDefectTypeEnum): The type of the defect.
+        cut_direction (List[int]): The direction of the cut as lattice vector, can be thought as a normal to the plane
+        that cuts the slab with added number of layers.
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/configuration.py:190]: This line needs to be indented

File: src/py/mat3ra/made/tools/build/defect/configuration.py [hunk 13]
```diff
@@ -146,20 +162,53 @@ class IslandSlabDefectConfiguration(SlabDefectConfiguration):
...
+        pivot_coordinate (List[float]): The pivot coordinate: the point in the unit cell
+        where the normal of the cut plane passes through.
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/configuration.py:192]: Same here

**Review by timurbazhirov — APPROVED** [review]

**Pull request merged**